qiskit/circuit/quantumcircuit.py

```diff
@@ -561,15 +561,15 @@ def _append(self, instruction, qargs, cargs):
     def _update_parameter_table(self, instruction):
         for param_index, param in enumerate(instruction.params):
             if isinstance(param, ParameterExpression):
-                current_parameters = self.parameters
+                current_parameters = self._parameter_table
 
                 for parameter in param.parameters:
                     if parameter in current_parameters:
                         if not self._check_dup_param_spec(self._parameter_table[parameter],
                                                           instruction, param_index):
                             self._parameter_table[parameter].append((instruction, param_index))
                     else:
-                        if parameter.name in {p.name for p in current_parameters}:
+                        if parameter.name in current_parameters:
```

Comment:
```suggestion
                        if parameter.name in {p.name for p in current_parameters}:
```
This needs to check `p.name` (the conflict it's looking for is separate `Parameters` with the same name), but the set comprehension can be moved outside the the two loops.

Reply:
I'm still a bit worried about the performance of this, even outside the loop we still have to iterate over the parameters each time we add a new one. That will get continually slower as the number of parameters grow. I'll push a fix up now, but I need to think if there is a way we can mitigate rebuilding the set on each new parameter

Reply:
A first pass at fixing this is in: https://github.com/Qiskit/qiskit-terra/pull/4232/commits/75d7349a05bdec16e8d4d249c975fea77a0184cb but I'm going to try and build the set iteratively somewhere (probably in ParameterTable too) to mitigate the performance hit of iterating over the dictionary on each gate add with a parameter.

Reply:
ok, removed the set comprehension and scaling limit in: https://github.com/Qiskit/qiskit-terra/pull/4232/commits/4913b44b03f327a28ccbd746885c15d423f78cbb